### PR TITLE
Conditional probabilities

### PR DESCRIPTION
**Pull request by L-v-M**
Improve report

### CODE REVIEW

File: shift_detector/checks/conditional_probabilities_check.py [hunk 1]
```diff
@@ -20,60 +84,105 @@ class ConditionalProbabilitiesCheck(Check):
...
         self.rule_limit = int(rule_limit)
         self.min_delta_supports = float(min_delta_supports)
         self.min_delta_confidences = float(min_delta_confidences)
+        self.number_of_bins = int(number_of_bins)
+        self.number_of_topics = int(number_of_topics)
 
     def run(self, store):
-        rules, examined_columns = store[
-            ConditionalProbabilitiesPrecalculation(self.min_support, self.min_confidence)]
+        pre_calculation = store[
```
Comment (velramiir) [inline shift_detector/checks/conditional_probabilities_check.py]: better to call it `precalculation_result`

File: shift_detector/checks/conditional_probabilities_check.py [hunk 2]
```diff
@@ -20,60 +84,105 @@ class ConditionalProbabilitiesCheck(Check):
...
+             abs(rule.delta_confidences) < self.min_delta_confidences
+             and abs(rule.delta_supports) >= self.min_delta_supports),
+            key=lambda r: (min(r.supports) != 0, abs(r.delta_confidences) * abs(r.delta_supports)),
+            reverse=True
+        )
+        add_to_explanation(orange_rules_falling_below_min_delta_confidences,
+                           'orange_rules_falling_below_min_delta_confidences')
 
-            explanation[', '.join(columns)].append(str(compressed_rule))
+        shifted_columns = set(tuple(sorted(attribute for attribute, _ in rule.left_side + rule.right_side))
```
Comment (velramiir) [inline shift_detector/checks/conditional_probabilities_check.py]: You don't need to convert this into a tuple if you convert into a set right after. just turn it into a set directly

File: docs/source/checks/conditional_probabilities.rst [hunk 3]
```diff
@@ -91,47 +91,61 @@ Parameters
...
+    of a right side (consequence) given a left side (antecedent) has to be at least 15%.
 
 ``rule_limit``:
-	This parameter expects an int and controls the maximum number of rules that are
-	printed as a result of executing :ref:`conditional_probabilities`.
-	The default value is 5. This means that the 5 most significant rules are printed.
+    This parameter expects an int and controls the maximum number of rules that are
+    printed in each section of the report as a result of executing
+    :ref:`conditional_probabilities`.
+    The default value is 5. This parameter does not have any impact on the diagram.
```
Comment (PaulZGit) [inline docs/source/checks/conditional_probabilities.rst]: vizualisation better than 'diagram'

File: docs/source/checks/conditional_probabilities.rst [hunk 4]
```diff
@@ -91,47 +91,61 @@ Parameters
...
-	This parameter expects a float between 0 and 1 and affects the granularity of the
-	comparison of the two data sets. Only rules whose support values exhibit an absolute
-	difference of more than ``min_delta_supports`` are considered during computation.
-	The default value is 0.05.
+    This parameter expects a float between 0 and 1 and affects the granularity of the
+    comparison of the two data sets. Only rules whose support values exhibit an absolute
+    difference of more than ``min_delta_supports`` are considered to indicate a shift.
+    A rule has to exceed ``min_delta_supports`` to be classified as an orange rule. If it
+    also exceeds ``min_delta_confidences`` it is classiefied as a red rule.
+    The default value is 0.05. This parameter does not impact the diagram.
```
Comment (PaulZGit) [inline docs/source/checks/conditional_probabilities.rst]: see comment above

File: docs/source/checks/conditional_probabilities.rst [hunk 5]
```diff
@@ -91,47 +91,61 @@ Parameters
...
 ``min_delta_confidences``:
-	This parameter expects a float between 0 and 1 and affects the granularity of the
-	comparison of the two data sets. Only rules whose confidence values exhibit an absolute
-	difference of more than ``min_delta_confidences`` are considered during computation.
-	The default value is 0.05.
+    This parameter expects a float between 0 and 1 and affects the granularity of the
+    comparison of the two data sets. Only rules whose confidence values exhibit an absolute
+    difference of more than ``min_delta_confidences`` are considered to indicate a shift.
+    A rule has to exceed ``min_delta_confidences`` to be classified as an orange rule. If it
+    also exceeds ``min_delta_supports`` it is classiefied as a red rule.
```
Comment (PaulZGit) [inline docs/source/checks/conditional_probabilities.rst]: typo in classiefied

File: docs/source/checks/conditional_probabilities.rst [hunk 6]
```diff
@@ -91,47 +91,61 @@ Parameters
...
+    The default value is 0.05. This parameter does not impact the diagram.
```
Comment (PaulZGit) [inline docs/source/checks/conditional_probabilities.rst]: vizualisation

File: docs/source/checks/conditional_probabilities.rst [hunk 7]
```diff
@@ -91,47 +91,61 @@ Parameters
...
+
+``number_of_bins``:
+    This parameter affects pre-processing of numerical columns.
+    Numerical columns are binned into ``number_of_bins`` many bins. The default value is 50.
+    This means that numerical columns are binned into 50 equal-width bins.
```
Comment (PaulZGit) [inline docs/source/checks/conditional_probabilities.rst]: equi-witdh

File: docs/source/checks/conditional_probabilities.rst [hunk 8]
```diff
@@ -147,33 +161,45 @@ Algorithm
...
+##############
+
+6. Intermediate result rules are partitioned in those rules that only appear in
+   the first data set, those rules that only appear in the second data set, and
+   those rules that appear in both data sets and consist of a left and a
+   right side (called mutual rules).
+7. Mutual rules are filtered for those exceeding ``min_delta_supports`` and
+   ``min_delta_confidences`` and sorted in descending order according to the
+   absolute difference of their confidence values and the maximum of their
+   supports of left side values. Rules whose support value is 0 in one data
```
Comment (PaulZGit) [inline docs/source/checks/conditional_probabilities.rst]: maybe briefly explain why they come last

File: docs/source/checks/conditional_probabilities.rst (hunk 7 above)
Comment (velramiir) [reply docs/source/checks/conditional_probabilities.rst]: I think equal-width is fine

File: shift_detector/checks/conditional_probabilities_check.py [hunk 9]
```diff
@@ -20,60 +84,105 @@ class ConditionalProbabilitiesCheck(Check):
...
+        to indicate a shift.
+    :param number_of_bins: an int greater than 0. Numerical columns are binned into ``number_of_bins``
+        bins.
+    :param number_of_topics: an int greater than 0. Textual columns are embedded into ``number_of_topics``
+        topics.
     """
 
     def __init__(self, min_support=0.01, min_confidence=0.15, rule_limit=5, min_delta_supports=0.05,
-                 min_delta_confidences=0.05):
+                 min_delta_confidences=0.05, number_of_bins=50, number_of_topics=20):
```
[the remark this answers is not in the document]
Comment (velramiir) [reply shift_detector/checks/conditional_probabilities_check.py]: You can also create an issue for this and do it later

File: docs/source/checks/conditional_probabilities.rst [hunk 10]
```diff
@@ -46,17 +46,17 @@ Code
...
-   :class:`~shift_detector.checks.ConditionalProbabilitiesCheck.ConditionalProbabilitiesCheck`.
-3. Finally, you print the result with :meth:`~shift_detector.Detector.Detector.evaluate`.
+   :class:`~shift_detector.checks.conditional_probabilities_check.ConditionalProbabilitiesCheck`.
+3. Finally, you print the result with :meth:`~shift_detector.detector.Detector.evaluate`.
 
 Result
 ++++++
 
-:ref:`conditional_probabilities` produces a set of so called rule clusters::
+:ref:`conditional_probabilities` produces a set of so called rule clusters and a summarizing diagram::
```
[the remark this answers is not in the document]
Comment (L-v-M) [reply docs/source/checks/conditional_probabilities.rst]: Yes, the documentation is not up to date yet

File: shift_detector/checks/conditional_probabilities_check.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (L-v-M) [reply shift_detector/checks/conditional_probabilities_check.py]: will do

File: shift_detector/checks/conditional_probabilities_check.py (hunk 1 above)
Comment (L-v-M) [reply shift_detector/checks/conditional_probabilities_check.py]: Good idea

File: shift_detector/checks/conditional_probabilities_check.py (hunk 2 above)
Comment (L-v-M) [reply shift_detector/checks/conditional_probabilities_check.py]: It's not a tuple, it's an iterator producing tuples.

File: docs/source/checks/conditional_probabilities.rst [hunk 11]
```diff
@@ -91,47 +91,61 @@ Parameters
 ----------
 
 :ref:`conditional_probabilities` provides several tuning knobs and adjustable
-thresholds that control (a) the computational complexity and
-(b) the size of the result:
+thresholds that control (a) the computational complexity,
```
[the remark this answers is not in the document]
Comment (L-v-M) [reply docs/source/checks/conditional_probabilities.rst]: I will update the documentation in another pr late. But will keep this in mind

File: docs/source/checks/conditional_probabilities.rst (hunk 4 above)
Comment (L-v-M) [reply docs/source/checks/conditional_probabilities.rst]: I will update the documentation in another pr late. But will keep this in mind

File: docs/source/checks/conditional_probabilities.rst (hunk 5 above)
Comment (L-v-M) [reply docs/source/checks/conditional_probabilities.rst]: I will update the documentation in another pr late. But will keep this in mind

File: docs/source/checks/conditional_probabilities.rst (hunk 6 above)
Comment (L-v-M) [reply docs/source/checks/conditional_probabilities.rst]: I will update the documentation in another pr late. But will keep this in mind

File: docs/source/checks/conditional_probabilities.rst (hunk 7 above)
Comment (L-v-M) [reply docs/source/checks/conditional_probabilities.rst]: I will update the documentation in another pr late. But will keep this in mind

File: docs/source/checks/conditional_probabilities.rst (hunk 8 above)
Comment (L-v-M) [reply docs/source/checks/conditional_probabilities.rst]: I will update the documentation in another pr late. But will keep this in mind

File: docs/source/checks/conditional_probabilities.rst (hunk 3 above)
Comment (L-v-M) [reply docs/source/checks/conditional_probabilities.rst]: I will update the documentation in another pr late. But will keep this in mind

**Pull request merged**